README.md
## Introduction

This is an asynchronous job scheduler with crontab syntax.

## How to use

Here is a basic example:

```python
# Should run as-is
import asyncio
from typing import Any
from datetime import datetime

from timewheel import TimeWheel
from timewheel.schedule import Schedule


async def my_job():
    print("hello from job!")
    await asyncio.sleep(3)


async def my_another_job(some_value: Any):
    print(f"Hey! This is my some_value {some_value}")
    await asyncio.sleep(1)

    
def my_special_job():
    print("Doing the special job!!")
    
async def main():
    timewheel = TimeWheel(schedules=[
        # Runs every 29 minutes using America/Sao_Paulo
        # as base
        Schedule(name="my-schedule", 
                 expression="*/29 * * * *",
                 timezone="America/Sao_Paulo",
                 job=my_job),
        # Runs every 5th, 10th and 20th minute on wednesday
        # using America/Los_Angeles tz as base
        Schedule(name="another-schedule", 
                 expression="5,10,20 * * * 2", 
                 timezone="America/Los_Angeles",
                 job=my_another_job),
        # Runs the special-job every minute except on 2021-12-31
        Schedule(name="special-job",
                 expression="* * * * *",
                 timezone="America/Sao_Paulo",
                 excluded_dates=[datetime.strptime('2021-12-31', '%Y-%m-%d')],
                 job=my_special_job),
        # Runs the jobs every minute between monday and friday 
        # using the range notation
        Schedule(name='another-job',
                 expression='* * * * 0-4',
                 timezone='America/Sao_Paulo',
                 job=my_special_job)
    ])
    await timewheel.run()


asyncio.get_event_loop().run_until_complete(main())
```

The timezone information is based on [IANA](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones).

CHANGELOG.md
## v.0.3.0 - 2022-05-25

**Added**

 - Now is possible to use the range notation in expression tokens.

## v.0.2.0 - 2021-12-28

**Added**

 - Now is possible to pass a list of dates which the job will not be executed.

## v.0.1.2 - 2021-12-13

**Added**

 - Additional validation in the `Scheduler` class to prevent duplicated executions.

## v.0.1.1 - 2021-12-10

**Changed**

 - Now is possible to set the verification frequency for running tasks using the `SCHEDULE_CHECK_INTERVAL`

## v.0.1.0 - 2021-04-29

**Added**

 - Now the `Schedule` class supports timezones.

**Changed**

 - The logger instances has been moved to inside the classes.

## v.0.0.5 - 2021-04-26

**Removed**

 - The signal handlers in order to block the uvicorn loop when working with fastapi.

## v.0.0.4 - 2021-04-25

**Fixed**

 - Now the schedule only runs if already finished the last execution.

## v0.0.3 - 2021-04-24

**Changed**

 - The method `_create_schedule` is a function now.

**Fixed**

 - When a job raises an exception the `Schedule` now logs the message instead of breaking.

## v0.0.2 - 2021-04-23

**Added**

 - Tests for the `Schedule` class

## v0.0.1 - 2021-04-23

**Added**

 - First working version

tests/test_sheduler.py
import logging

import pytest
from datetime import datetime
from unittest.mock import Mock, AsyncMock

from timewheel.schedule import Schedule, EXPRESSION_VALIDATOR_REGEXP


# Instance creation tests

def test_create_schedule_running_every_minute():
    s = Schedule(name='some-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=print)

    assert s.name == 'some-schedule'
    assert s.schedule_table.minutes
    assert s.schedule_table.hours
    assert s.schedule_table.days
    assert s.schedule_table.months
    assert s.schedule_table.weekdays


def test_create_schedule_with_nth_expression():
    s = Schedule(name='some-schedule',
                 expression='*/2 * * * *',
                 timezone='America/Sao_Paulo',
                 job=print)

    assert isinstance(s.schedule_table.minutes, list)
    assert len(s.schedule_table.minutes) == 30


def test_create_schedule_with_multi_tokens():
    s = Schedule(name='some-schedule',
                 expression='1,2,3,4,5 * * * *',
                 timezone='America/Sao_Paulo',
                 job=print)

    assert isinstance(s.schedule_table.minutes, list)
    assert len(s.schedule_table.minutes) == 5


def test_create_schedule_with_range_token():
    s = Schedule(name='my-range-schedule',
                 expression='* * * * 1-5',
                 timezone='America/Sao_Paulo',
                 job=print)

    assert isinstance(s.schedule_table.weekdays, list)
    assert len(s.schedule_table.weekdays) == 5


def test_create_schedule_with_complex_expression():
    s = Schedule(name='my-complex-schedule',
                 expression="*/2 20 1 * 0,3-6",
                 timezone='America/Sao_Paulo',
                 job=print)

    assert isinstance(s.schedule_table.minutes, list) and len(s.schedule_table.minutes) == 30
    assert isinstance(s.schedule_table.weekdays, list) and len(s.schedule_table.weekdays) == 5


def test_create_schedule_with_nth_token_having_value_over_the_max():
    expression = '*/100 * * * *'
    with pytest.raises(ValueError) as error:
        Schedule(name='some-schedule',
                 expression=expression,
                 timezone='America/Sao_Paulo',
                 job=print)

    assert f"The expression {expression} contains an invalid token. " \
           f"Details: The nth token value 100 is higher then the " \
           f"max value (59) for the field type." == str(error.value)


def test_create_schedule_with_token_having_value_over_the_max():
    expression = '100 * * * *'
    with pytest.raises(ValueError) as error:
        Schedule(name='some-schedule',
                 expression=expression,
                 timezone='America/Sao_Paulo',
                 job=print)

    assert f"The expression {expression} contains an invalid token. " \
           f"Details: The token value 100 is higher then the " \
           f"max value (59) for the field type." == str(error.value)


def test_create_schedule_with_multi_token_and_all():
    expression = '1,2,3,4,* * * * *'

    with pytest.raises(ValueError) as error:
        Schedule(name='some-schedule',
                 expression=expression,
                 timezone='America/Sao_Paulo',
                 job=print)

    assert f"The expression {expression} contains an invalid token. " \
           f"Details: Invalid syntax for the token 1,2,3,4,*" == str(error.value)


def test_create_schedule_with_invalid_expression():

    with pytest.raises(ValueError) as error:
        Schedule(name='some-schedule',
                 expression='* * * *',
                 timezone='America/Sao_Paulo',
                 job=print)

    assert f"The expression must match the pattern '{EXPRESSION_VALIDATOR_REGEXP.pattern}'" == str(error.value)


# Schedule validation

def test_should_run_true():
    current_time = datetime.strptime('2021-10-01 00:00:00', '%Y-%m-%d %H:%M:%S')
    s = Schedule(name='my-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=print)
    assert s.schedule_table.should_run(current_time)


def test_should_run_false():
    current_time = datetime.strptime('2021-10-01 00:05:00', '%Y-%m-%d %H:%M:%S')
    s = Schedule(name='my-schedule',
                 expression='*/7 * * * *',
                 timezone='America/Sao_Paulo',
                 job=print)

    assert not s.schedule_table.should_run(current_time)


@pytest.mark.asyncio
async def test_finish_scheduler():
    s = Schedule(name='my-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=print)

    await s.finish()
    assert s.stop_signal_received
    assert not s.running


@pytest.mark.asyncio
async def test_run_with_stop_signal_received():
    s = Schedule(name='my-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=print)
    s.stop_signal_received = True

    r = await s.run()
    assert not r


@pytest.mark.asyncio
async def test_run_with_excluded_date_set(caplog):

    excluded_dates = [
        datetime.today().date()
    ]

    job = Mock()

    s = Schedule(name='my-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=job,
                 excluded_dates=excluded_dates)
    with caplog.at_level(logging.INFO):
        await s.run()

        job.assert_not_called()
        assert "has been added to the exclusion list" in caplog.text


@pytest.mark.asyncio
async def test_run_sync_job_successfully(caplog):

    job = Mock()

    s = Schedule(name='my-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=job)
    with caplog.at_level(logging.INFO):
        await s.run()

        job.assert_called_once()
        assert "Running the schedule with the job" in caplog.text


@pytest.mark.asyncio
async def test_run_sync_job_ensure_running_once_per_minute(caplog):
    job = Mock()

    s = Schedule(name='my-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=job)
    with caplog.at_level(logging.INFO):
        for _ in range(10):
            await s.run()

        job.assert_called_once()
        assert "Running the schedule with the job" in caplog.text


@pytest.mark.asyncio
async def test_run_coro_job_successfully(caplog):
    job = AsyncMock()

    s = Schedule(name='my-schedule',
                 expression='* * * * *',
                 timezone='America/Sao_Paulo',
                 job=job)

    with caplog.at_level(logging.INFO):
        await s.run()

        job.assert_called_once()
        assert "Running the schedule with the job" in caplog.text



tests/test_timewheel.py
import asyncio
import logging
from unittest.mock import Mock

import pytest

from timewheel.schedule import Schedule
from timewheel.wheel import TimeWheel


def test_setup_timewheel_with_valid_parameters():
    my_tw = TimeWheel(schedules=[])

    assert isinstance(my_tw, TimeWheel)
    assert my_tw.running is True


@pytest.mark.asyncio
async def test_timewheel_runner(caplog):

    job = Mock(return_value=None)

    my_tw = TimeWheel(schedules=[Schedule(name='my-job',
                                          expression='* * * * *',
                                          timezone='America/Sao_Paulo',
                                          job=job)])

    with caplog.at_level(logging.INFO):
        t = asyncio.create_task(my_tw.run())
        await asyncio.sleep(.1)
        my_tw.running = False
        await asyncio.sleep(1)

        job.assert_called()
        assert "Finished the timewheel loop" in caplog.text
        t.cancel()




timewheel/wheel.py
import os
import logging
import asyncio
from typing import List

from timewheel.schedule import Schedule

SLEEP_TIME = 1
SCHEDULE_CHECK_INTERVAL = os.getenv('SCHEDULE_CHECK_INTERVAL', '10')


class TimeWheel:

    def __init__(self, schedules: List[Schedule]):
        try:
            self.logger = logging.getLogger("timewheel")
            self.schedules = schedules
            self.schedule_check_interval = int(SCHEDULE_CHECK_INTERVAL)
            self.running = True
        except ValueError:
            raise ValueError(f"The variable SCHEDULE_CHECK_INTERVAL must be an integer, "
                             f"received value {SCHEDULE_CHECK_INTERVAL}")

    async def run(self):
        """
        Starts the scheduler loops

        :return:
        """
        timer = 0

        while self.running:
            # Only check the schedule every SCHEDULE_CHECK_INTERVAL seconds
            if timer % self.schedule_check_interval == 0:
                timer = 0
                for schedule in self.schedules:
                    if not schedule.running:
                        asyncio.create_task(schedule.run())
            await asyncio.sleep(SLEEP_TIME)
            timer += SLEEP_TIME
        self.logger.warning("Finished the timewheel loop!")

    async def kill_jobs(self):
        """
            When the application receives a system signal to terminate
            invokes the finish methods for all schedules in order to
            wait them finish the jobs if there are any running.
        """

        self.logger.warning("Received system signal to finish!")
        for schedule in self.schedules:
            self.logger.debug(f"Waiting for the schedule {schedule.name} to finish...")
            await schedule.finish()
        self.running = False


timewheel/schedule.py
import re
import time
import asyncio
import logging
from datetime import datetime, date
from zoneinfo import ZoneInfo
from dataclasses import dataclass
from typing import Callable, List, Union


# This regular expression is used to validate the crontab expression
# passed to the constructor.
EXPRESSION_VALIDATOR_REGEXP = re.compile(r"((([\d*]+|[\d*]/\d|\d-\d),?)+ ?){5}")
EXPRESSION_SEPARATOR = " "
EXPECTED_TOKENS = 5
ALL_TOKEN = "*"
NTH_TOKEN = "/"
MULTI_TOKEN = ","
RANGE_TOKEN = "-"


@dataclass
class ScheduleTable:
    minutes: Union[List[int], bool]
    hours: Union[List[int], bool]
    days: Union[List[int], bool]
    months: Union[List[int], bool]
    weekdays: Union[List[int], bool]

    def should_run(self, current_time: datetime) -> bool:
        """
        Check if the current_timestamp is true for all tokens and return

        :param current_time: The current timestamp received from the external loop
        :return: Boolean informing if the action should be called
        """
        return all([self.minutes is True or current_time.minute in self.minutes,
                    self.hours is True or current_time.hour in self.hours,
                    self.days is True or current_time.day in self.days,
                    self.months is True or current_time.month in self.months,
                    self.weekdays is True or current_time.weekday() in self.weekdays])


class Schedule:
    """
        Class responsible for create the schedule
        table and check if the action should be run.
    """

    def __init__(self,
                 name: str,
                 expression: str,
                 timezone: str,
                 job: Callable,
                 excluded_dates: List[date] = None):
        """
        :param name: The schedule name
        :param expression: A crontab expression, for more information
            please check: https://en.wikipedia.org/wiki/Cron
        :param timezone: Timezone name based on IANA database. Please
            check: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
        :param excluded_dates: List containing the date which the schedule
            will not run.
        :param job: An awaitable object to be run.
        """
        self.logger = logging.getLogger('timewheel.scheduler')
        self.name = name
        self.schedule_table = create_schedules(expression)
        self.timezone = ZoneInfo(timezone)
        self.excluded_dates = excluded_dates
        self.job = job
        self.running = False
        self.last_execution = None
        self.stop_signal_received = False

        self.logger.info(f"[{self.name}]: Created scheduler for the expression "
                         f"{expression} and job {job} using the timezone {timezone}")

        self.logger.debug(f"[{self.name}]: Schedule table data: {self.schedule_table}")

    async def run(self):
        """
        Checks if the job should be run in the current timestamp
        for the schedule timezone.
        If the scheduler received the signal to finish, than the
        execution is ignored.

        Important: The Callable should never return any value or
        the execution will be marked as error.

        :return: None
        """
        current_time = datetime.now(tz=self.timezone).replace(second=0, microsecond=0)
        self.logger.debug(f"[{self.name}]: Checking if should run at {current_time} using the tz {self.timezone}")

        if self.stop_signal_received:
            return

        if self.schedule_table.should_run(current_time):
            # Returns if the current_date is set as excluded
            if self.excluded_dates is not None and current_time.date() in self.excluded_dates:
                self.logger.info(f"[{self.name}]: The {self.job} should be executed, but the date "
                                 f"{current_time.date()} has been added to the exclusion list.")
                return
            # This check is used to avoid the execution of a task more then once in
            # a minute
            if self.last_execution is None or current_time != self.last_execution:
                self.last_execution = current_time
                self.running = True
                succeeded = True
                start = time.time()
                self.logger.info(f"[{self.name}]: Running the schedule with the job {self.job}")
                if asyncio.iscoroutinefunction(self.job):
                    result = await asyncio.gather(self.job(),
                                                  return_exceptions=True)
                else:
                    result = await asyncio.gather(
                        asyncio.get_running_loop().run_in_executor(None, self.job),
                        return_exceptions=True
                    )
                if result and result[0] is not None:
                    succeeded = False
                    self.logger.error(f"[{self.name}]: An unexpected error occurred "
                                      f"while running the job {self.job}. Details: "
                                      f"{result}")
                self.logger.info(f"[{self.name}]: Job finished in {time.time() - start} seconds with status {succeeded}.")
                self.running = False

    async def finish(self):
        """
        Set the stop signal as True in order to avoid
        new executions and waits the instance finish
        the action execution.
        """
        self.logger.warning(f"{self.name}: Finish signal received")

        self.stop_signal_received = True
        while self.running:
            await asyncio.sleep(0.5)


def create_schedules(expression: str) -> ScheduleTable:
    """
    Convert a cron expression into a ScheduleTable.

    The expression structure is the following:

     ┌───────────── minute (0 - 59)
     │ ┌───────────── hour (0 - 23)
     │ │ ┌───────────── day of the month (1 - 31)
     │ │ │ ┌───────────── month (1 - 12)
     │ │ │ │ ┌───────────── day of the week (0 - 6)
     │ │ │ │ │
     * * * * *

    :param expression: String containing the cron expression
    :return:
    """
    if not re.fullmatch(EXPRESSION_VALIDATOR_REGEXP, expression):
        raise ValueError(f"The expression must match the pattern "
                         f"'{EXPRESSION_VALIDATOR_REGEXP.pattern}'")

    tokens = expression.split(EXPRESSION_SEPARATOR)
    try:
        return ScheduleTable(minutes=parse_expression_token(tokens[0], 0, 59),
                             hours=parse_expression_token(tokens[1], 0, 23),
                             days=parse_expression_token(tokens[2], 1, 31),
                             months=parse_expression_token(tokens[3], 1, 12),
                             weekdays=parse_expression_token(tokens[4], 0, 6))

    except ValueError as error:
        raise ValueError(f"The expression {expression} contains an "
                         f"invalid token. Details: {''.join(error.args)}") from error


def parse_expression_token(token: str,
                           start_value: int,
                           max_value: int) -> Union[List[int], bool]:
    """
    Converts a token into a list of integers or a bool

    :param token: The string containing the token expression. Eg: */2,13
    :param start_value: The initial value used to create the range
    :param max_value: The max value allowed for the token. This value is
        also used to generate the values list.
    :return:
        If the token value is an asterisk returns True
        else returns a list containing the integer values
    """
    extracted_tokens = token.split(MULTI_TOKEN)

    if len(extracted_tokens) > 1 and ALL_TOKEN in extracted_tokens:
        raise ValueError(f"Invalid syntax for the token {token}")

    generated_tokens = []
    for extracted_token in extracted_tokens:
        if extracted_token == ALL_TOKEN:
            return True
        if NTH_TOKEN in extracted_token:
            generated_tokens.extend(parse_nth_token(extracted_token,
                                                    start_value,
                                                    max_value))
        elif RANGE_TOKEN in extracted_token:
            generated_tokens.extend(parse_range_token(extracted_token, start_value, max_value))
        else:
            extracted_token = int(extracted_token)
            if extracted_token > max_value:
                raise ValueError(f"The token value {extracted_token} is higher "
                                 f"then the max value ({max_value}) for the "
                                 f"field type.")
            generated_tokens.append(extracted_token)

    return sorted(set(generated_tokens))


def parse_range_token(range_token: str,
                      start_value: int,
                      max_value: int) -> List[int]:
    """
    Parses a range expression into a range
    of values between the start_value and
    max_value

    :param range_token: The token containing the range pattern X-Y
    :param start_value: The minimal value for the token
    :param max_value: The max value for the token
    :return: A list containing the integers for the
        specified range.
    """
    if range_token.count(RANGE_TOKEN) > 1:
        raise ValueError(f"The range notation expects the pattern X-Y, but the "
                         f"token is {range_token}")

    x, y = map(int, range_token.split(RANGE_TOKEN))

    if x > y:
        raise ValueError(f"The initial value must be less or equal to the end value")

    if x < start_value or y > max_value:
        raise ValueError(f"The values for the range must be between {start_value} and {max_value}")

    return list(range(x, y + 1))


def parse_nth_token(nth_token: str,
                    start_value: int,
                    max_value: int) -> List[int]:
    """
    Parses a nth token expression and generates
    the corresponding list of integers.

    :param nth_token: nth expression. E.g: */2
    :param start_value: Initial value for the interval
    :param max_value: Max value for the interval
    :return: A list containing all values which the
        mod from the right value is zero
    """
    nth = int(nth_token.split(NTH_TOKEN)[1])
    if nth > max_value:
        raise ValueError(f"The nth token value {nth} is higher then the "
                         f"max value ({max_value}) for the field type.")

    return([value for value in range(start_value, max_value + 1)
            if value % nth == 0])
